Orange/data/variable.py

```diff
@@ -887,114 +635,121 @@ class TimeVariable(ContinuousVariable):
-                return float(datestr)
-            except ValueError:
-                raise ERROR
+        Parameters
+        ----------
+        column : pd.Series
+            The column to transform.
 
-        for i, (have_date, have_time, fmt) in enumerate(self._ISO_FORMATS):
-            try:
-                dt = datetime.strptime(datestr, fmt)
-            except ValueError:
+        Returns
+        -------
+        A transformed pd.Series of the datetime type, with timezone information.
+        """
+        for val in column:
+            # handle missing values like they don't exist
+            if val in Variable.MISSING_VALUES or (isinstance(val, Number) and np.isnan(val)):
                 continue
-            else:
-                # Pop this most-recently-used format to front
-                if 0 < i < len(self._ISO_FORMATS) - 2:
-                    self._ISO_FORMATS[i], self._ISO_FORMATS[0] = \
-                        self._ISO_FORMATS[0], self._ISO_FORMATS[i]
-
-                self.have_date |= have_date
-                self.have_time |= have_time
-                if not have_date:
-                    dt = dt.replace(self.UNIX_EPOCH.year,
-                                    self.UNIX_EPOCH.month,
-                                    self.UNIX_EPOCH.day)
+            # for multiple timezones, use the last one for display
+            self.timezone = TimeVariable._detect_timezone(val) if not np.issubdtype(column.dtype, np.number) else None
+            # if any value doesn't have a timezone, permanently strip display timezones for the column
+            if self.timezone is None:
                 break
+
+        # if the columns are integers (timestamps), use different logic
+        # than when we are dealing with strings
+        if 'format' not in self.attributes and np.issubdtype(column.dtype, np.number):
+            # timestamps are seconds
+            kwargs = {'unit': 's'}
+        else:
+            # allow the variable to specify a format (overrides integers)
+            kwargs = {'format': self.attributes.get('format')}
+
+        # .apply(str) because parsing a float discards the fractional part for some reason
+        # utc=True: make timezone aware
+        # .values: return a DatetimeIndex so we can actually localize to UTC
+        result = pd.to_datetime(column.apply(str).values, errors='raise', exact=True, utc=True,
+                                infer_datetime_format=True, **kwargs)
+
+        # determine whether we should display the time part (HH:MM:SS.MS)
+        # only display it when hours, minutes and seconds are all 0 in all cases
+        # this is the most robust way as it doesn't depend on regexes
+        # (we might not even have strings as inputs) and offloads any parsing to pandas
+        # addition works because everything is non-negative
+        self.has_time_component = (result.hour + result.minute + result.second + result.microsecond).sum() != 0
```

Comment:
Could we instead use `pd.tseries.tools._guess_datetime_format()` and save a format string, like `"%Y"`. Sure, it's a private interface, but we can catch that in a try-except close at the top of the module, and just adapt it when it changes.
It would be _so much nicer_ to format datetimes constructed from yearly series as `"%Y"` only ...